### ISSUE TITLE
Package version conflicts do not lead to error

### ISSUE THREAD

**Issue by stevn** [conan-io/conan#17397]
I think this is similar to #17190.

I am still slightly confused about the behaviour of Conan 2 and I currently don't have time to investigate this in more detail.

It seems that the depender package sometimes (always?) automatically overrides versions of the requires() of a dependency. In some cases that is OK. In other cases that is fatally flawed and the depender package doesn't notice that something bad is happening until something breaks in production.

I would like to get at least a warning about such overrides of versions. And I would like to be able to make that warning an error.

Basically this concept should apply to everything that is overridden between recipes anywhere.

Every recipe was written with a certain intention. If that intention is then completely being ignored when something doesn't fit together, I need to be able to catch that before bad things happen. I want to resolve conflicts myself because only then I can build a stable and production-ready package ecosystem.

**Comment by jcar87** [comment]
Hi @stevn  - thanks for opening this issue.

> It seems that the depender package sometimes (always?) automatically overrides versions of the requires() of a dependency.

The Conan 2 solver should error with a conflict when different versions of the same requirement exists in multiple places, as per the docs: https://docs.conan.io/2/tutorial/versioning/conflicts.html

Only when `override=True` or `force=True` is used would you see a different behaviour - howewer, this is only meant as a temporary solution for testing something locally (and overrides should be used in the top-level consumer), and not something "baked" into the recipes.

As per the docs:

> Resolving version conflicts by overrides/forces should in general be the exception and avoided when possible, applied as a temporary workaround. The real solution is to move forward the dependencies requires so they naturally converge to the same versions of upstream dependencies.

This last line in the docs seem to align with what you suggest:

>  I want to resolve conflicts myself because only then I can build a stable and production-ready package ecosystem.

Is it possible that you are working with recipes that have `override` in recipes along the graph? We do not allow overrides on Conan Center precisely to avoid this very problem. 


> I would like to get at least a warning about such overrides of versions. And I would like to be able to make that warning an error.

When printing the resolved dependencies, the summary should have a section that lists the overrides - for example

```
======== Computing dependency graph ========
Graph root
    conanfile.py 
Requirements
    argparse/2.9#672db1e97573ed9df006df4cfcd6e049 - Cache
    benchmark/1.8.0#ff63873550d1cb2efc64b2785837c23a - Cache
    boost/1.84.0#6815195f2bf512487e7beb7e43ca31b6 - Cache
   [... clipped for brevity ...]
Build requirements
    autoconf/2.71#5312af4e8a3f82a7f6454866081de83f - Cache
    automake/1.16.5#27a1996f8e84efd37c1bdacd9c8a5d7b - Cache
   [... clipped for brevity ...]
Overrides
    fmt/10.1.1: ['fmt/[10.0.0]']
    boost/1.83.0: ['boost/1.84.0']
```

Where this `boost/1.83.0: ['boost/1.84.0']` indicates that a requirement on `boost/1.83.0` was overridden and satisfied by `boost/1.84.0` instead. This is also part of the json output, which can be programmatically filtered to catch these if necessary.

**Comment by memsharded** [comment]
Hi @stevn 

Did the above helped? In Conan 2 package version conflicts are always an error by default, unless explicitly defined.

**Comment by stevn** [comment]
@memsharded Thanks for following up on this! I am so sorry, but I still couldn't find the time to build a reproduction scenario. The project is a bit complex and I still don't quite understand what exactly happened.

Anyway, I saw this warning message in the build logs of a **Conan v1** project, which seemed very weird to me.

```
WARN: file_io/4.2.0: requirement file_types/[^2.0.0] overridden by file_mux/8.0.1 to file_types/1.1.0
```

The version of file_types being used (v1.1.0) is lower than the specified version range allows (^2.0.0). So I would have expected this to not be possible.

Gladly, in this case the compilation actually failed because file_io was using new definitions that were only introduced later - in file_types v2.

I would have expected the package manager to catch this early. Instead of having to read the compiler errors, **I would have expected a package version conflict error message**.

In this case it was not a huge issue because the compile failed. This kind of thing can be a real problem if the compile and build works OK and the whole thing blows up at customer site later because it might result in undefined behavior in some cases I guess. There are other ways to catch those kinds of things, but still - I am just concerned that this version conflict isn't caught and then gets shipped accidentally.

I don't know if the **behavior of Conan v2** changed in this regard? Is this still possible? Or are you sure this is impossible to happen in Conan v2?

If needed, I can maybe try to build a reproduction case next week. Maybe that would be helpful to confirm a fix in Conan v2?

**Comment by memsharded** [comment]
> I would have expected the package manager to catch this early. Instead of having to read the compiler errors, I would have expected a package version conflict error message.

Yes, Conan 2 is expected to not override by default, and it will give an error message of conflict instead.

> I don't know if the behavior of Conan v2 changed in this regard? Is this still possible? Or are you sure this is impossible to happen in Conan v2?

Yes, Conan 2 changed this. Now Conan 2 will always raise errors for any conflict. Unless recipes explicitly add ``override=True`` or ``force=True``, all version conflicts are considered an error and they will raise. If overrides happen, they will also be summarized in the commands output, it should be very visible that such override has happened.

Sure, a reproduction case would be very helpful, thanks for the feedback!

**Comment by stevn** [comment]
@memsharded @jcar87 

Thanks for your feedback! I have now reproduced and analyzed the dependencies situation that caused this problem. The reproduction project is available at:

https://github.com/stevn/conan_version_conflict_test

I was able to confirm that the error handling for this type of package version conflict is fixed in Conan v2!

So I think there is no further action really required. I am glad to see this improvement in Conan v2 and look forward to fully switching over to the latest version.

**Comment by memsharded** [comment]
> I was able to confirm that the error handling for this type of package version conflict is fixed in Conan v2!

Excellent, happy to heat that!

> So I think there is no further action really required. I am glad to see this improvement in Conan v2 and look forward to fully switching over to the latest version.

Thanks very much for following up and your feedback
